components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java

```diff
@@ -1093,27 +1099,39 @@ private void handleAuthenticationReponseFromFramework(HttpServletRequest req, Ht
-
-                        sendNotification(errorResp, SAMLSSOConstants.Notification.EXCEPTION_STATUS, SAMLSSOConstants
-                                .Notification.EXCEPTION_MESSAGE, assertionConsumerURL, req, resp);
-                        return;
-                    } else {
+                    String authDestinationUrl = authnReqDTO.getDestination();
+                    if (idpDestinationURLs.contains(authDestinationUrl)) {
                         if (log.isDebugEnabled()) {
-                            log.debug("Successfully validated destination of the authentication request of issuer :"
-                                    + issuer + " in tenant domain : " + tenantDomain);
+                            log.debug("Successfully validated destination of the authentication request " +
+                                    "of issuer :" + issuer + " in tenant domain : " + tenantDomain);
+                        }
+                    } else {
+                        try {
+                            URL destinationUrl = new URL(authnReqDTO.getDestination());
+                            if (destinationUrl.getProtocol().equals(HTTPS_SCHEME) && destinationUrl.getPort() == -1) {
+                                authDestinationUrl = new URL(destinationUrl.getProtocol(), destinationUrl
+                                        .getHost(), DEFAULT_HTTPS_PORT, destinationUrl.getFile()).toString();
+                            } else if (destinationUrl.getProtocol().equals(HTTP_SCHEME) && destinationUrl
+                                    .getPort() == -1) {
+                                authDestinationUrl = new URL(destinationUrl.getProtocol(), destinationUrl
+                                        .getHost(), DEFAULT_HTTP_PORT, destinationUrl.getFile()).toString();
+                            }
+                        } catch (MalformedURLException e) {
+                            // This block is reached if the destination url is a relative url. Since spec doesn't
+                            // restrict this exception will not be handled.
+                        }
+                        if (idpDestinationURLs.contains(authDestinationUrl)) {
+                            if (log.isDebugEnabled()) {
+                                log.debug("Successfully validated destination of the authentication request " +
+                                        "of issuer :" + issuer + " in tenant domain : " + tenantDomain);
+                            }
+                        } else {
+                            String msg = "Destination validation for authentication request failed. " + "Received: " +
+                                    authDestinationUrl + "." + " Expected one in the list: [" + StringUtils
+                                    .join(idpDestinationURLs, ',') + "]";
+                            handleInvalidRequest(msg, authnReqDTO, req, resp);
+                            return;
```

Comment:
Shall we move this validation to separate method, so that it is more readable and independently unit testable? The code looks fine to me.